### ISSUE TITLE
Proposal: support v1 and v2 mapping

### ISSUE THREAD

**Issue by lmihalkovic** [commanderx16/x16-emulator#327]
The idea is to allow people to test/migrate their code before the board is there

In my code I did (final code is not exactly this, but close enough to understand the concept)

```
	devices_v1[0] = io_page_create(0x9f00, 0x20, NULL, NULL);		// sound
	devices_v1[1] = io_page_create(0x9f20, 0x20, NULL, NULL);	        // VERA
	devices_v1[2] = io_page_create(0x9f40, 0x20, NULL, NULL);		// lcf screen
	devices_v1[3] = io_page_create(0x9f60, 0x20, NULL, NULL);		// VIA
	devices_v1[4] = io_page_create(0x9f80, 0x20, NULL, NULL);		// bus 1
	devices_v1[5] = io_page_create(0x9fa0, 0x20, NULL, NULL);		// bus 2
	devices_v1[6] = io_page_create(0x9fc0, 0x20, NULL, NULL);		// bus 3
	devices_v1[7] = io_page_create(0x9fe0, 0x20, NULL, NULL);		// bus 4


	devices_v2[0] = io_page_create(0x9f00, 0x20, NULL, NULL);		// VIA
	devices_v2[1] = io_page_create(0x9f20, 0x20, NULL, NULL);		// VERA
	devices_v2[2] = io_page_create(0x9f40, 0x20, NULL, NULL);		// sound
	devices_v2[3] = io_page_create(0x9f60, 0x20, NULL, NULL);		// bus 1
	devices_v2[4] = io_page_create(0x9f80, 0x20, NULL, NULL);		// bus 2
	devices_v2[5] = io_page_create(0x9fa0, 0x20, NULL, NULL);		// bus 3
	devices_v2[6] = io_page_create(0x9fc0, 0x20, NULL, NULL);		// bus 4
	devices_v2[7] = io_page_create(0x9fe0, 0x20, NULL, NULL);		// bus 5  
```

Based on this model, it is also possible to create dynamic libs that can provide emulator support for the external hardware that people will create... also not difficult to extend the model to load the hardware map from a JSON file.

Is there enough interest to justify making a PR for it? (all the code is in a single file)

**Comment by Elektron72** [comment]
Although this would be useful, it is likely that the maintainers of the repository won't return until after the next board revision is completed, making a pull request that supports both v1 and v2 obsolete.

**Comment by lmihalkovic** [comment]
if you have any way to ping them, could you do that for me please?!  np ... maybe I will just make my repo public... 

the real idea behind what I am doing is not explained, but bottom line I do need to be able to quickly re-define the IO area.

so support for v1 and v2 in the same emulator instance (as opposed to the v2 branch) is just a quick way to test my code. another 'free' side-effect is being able to extend the emulator for the people who want to make custom HW boards... without having to change the emulator source code (just define their own dynamic lib).

I also cleaned up the memory/bank code... makes it trivial to experiment with different memory layouts (I am making a small HW proto that does not have much mem).

**Comment by lmihalkovic** [comment]
so this is the real code

```
    io_page_t *v1[11] = {
        io_page_create(0x9f00, REGS_32, "sound", NULL, NULL, NULL),			
        io_page_create(0x9f20, REGS_32, "VERA",  video_write, NULL, video_read),
        io_page_create(0x9f40, REGS_32, "lcd",   NULL, NULL, NULL),	
        io_page_create(0x9f60, REGS_16, "VIA1",  via1_write, via1_read, NULL),
        io_page_create(0x9f70, REGS_16, "VIA2",  via2_write, via2_read, NULL),
```

**Comment by mist64** [comment]
I agree with @Elektron72 in that the switch to Proto #2 has to be a hard one, without a way of going back. In fact, I already merged the changes for Proto #2; all future builds won't support Proto #1 any more.

As for your design suggestion in general: I'm not sure. MAME/MESS does it similarly to what you're suggesting, because they need to be very flexible: Different devices reuse the same components, so generic data structures define the memory layout.

For the X16, this seems like overkill to me. There's a long way until a handful of #ifdefs won't accomplish what you need – with improved readability up to some point.

**Comment by lmihalkovic** [comment]
I did all these things in my codebase ... it took 4h to get it right because I haven't touched C in a little while. I think it cleans up the code quite a bit, but that is just a matter of personal preference/perspective. 

my motivation was just to clean things... and my own 6502 proto does not have the same memory layout... so being able to play with different configs was in my mind useful, considering how trivial it was to do (you code is very well organized, making what I did trivial todo). the second motivation was VERA ... and running 2 cards in one computer (having the debugger window opened while some code runs on a different window is kinda cool... so I thought.
